docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,170 @@
+This IOCTL emulates the mprotect syscall with SGX specific extensions. In future kernel implementation, it could be mprotect or pkey_mprotect syscall with sgx extensions for the "prot" parameter.
+
+```
+#define SGX_IOC_ENCLAVE_MPROTECT _IOW(SGX_MAGIC, 0x06, struct sgx_enclave_mprotect)
+/**
+ * struct sgx_enclave_mprotect - parameter structure for the
+ *                                %SGX_IOC_ENCLAVE_MPROTECT ioctl
+ * @addr:	address of the memory to change protections
+ * @length:	length of the area.  This must be a multiple of the page size.
+ * @prot:	this must be or'ed of following:
+             PROT_READ
+             PROT_WRITE
+             PROT_EXEC
+             PROT_TRIM (new): change the page type to PT_TRIM, implies RW. User space should immediately EACCEPT, and then  call mprotect with PROT_NONE.
+             PROT_TCS (new): change the page type to PT_TCS
+	     PROT_NONE: Signal the kernel EACCEPT is done for PT_TRIM pages. Kernel can EREMOVE the pages at a time it deems appropriate.
+ */
+struct sgx_enclave_mprotect {
+	__u64	addr;
+	__u64	length;
+	__u64	prot;
+};
+```
+
+### Remarks
+
+Kernel should ensure that SGX instructions can succeed or catch and handle any fault.
+  - The kernel may maintain EPCM information on each page which includes access permission RWX, page types of PT_REG, PT_TRIM, PT_TCS.
+  - The kernel should EREMOVE pages of PT_TRIM only after user space signals kernel EACCEPT is done with mprotect(...,PROT_NONE,...). This is because EACCEPT may cause undesired #PF if the target page is already EREMOVED.
+  - The kernel catches fault on EMODPR, EMODT and converts to error code returned to user space.
+
+The enclave run-time (or trusted run-time) may implement a parallel memory management structure which would provide information to the enclave on the enclave memory mappings.  The run-time can have a trusted API analogous to mmap which makes a call out of the enclave to issue the mmap  and then either perform EACCEPT on the pages and update the internal memory structures or configure the enclave to perform the EACCEPT when a #PF is delivered to the enclave.  With the outlined kernel interface, either implementation is possible.
+## Sequence Diagrams for Basic Flows
+
+### Direct Allocation with MAP_POPULATE
+
+![SGX2 direct allocation flow](images/SGX2-alloc-direct.svg)
+
+### \#PF Based Allocation
+
+![SGX2 #PF based allocation flow](images/SGX2-alloc-pf.svg)
+
+### EPC Deallocation
+
+![SGX2 deallocation flow](images/SGX2-trim.svg)
+
+### Permission Changes
+
+![SGX2 permissions change flow](images/SGX2-perms.svg)
+
+**Notes:**
+- EACCEPT is needed for enclave to ensure untrusted runtime and OS indeed invoke EMODPR and EPCM permissions are set as expected.
+- It is assumed that both OS and enclave keep track of page permissions. However, it is possible to avoid that as the current [Intel SDK implementation](https://github.com/intel/linux-sgx/blob/master/sdk/trts/trts_ecall.cpp#L561) demostrates.
```

Comment:
Can you elaborate on the premise on why both the OS and enclave would be keeping track of page permissions, as well as why the existing Intel SGX SDK implementation elides this? It seems to hinge on the partitioning of responsibilities between the OS for EMODPR for restrictions and enclave for EMODPE requiring EACCEPT, but it's not clear that the enclave is not tracking page permissions in that case.

Reply:
For OS to know when to do EMODPR, it needs to keep track of EPCM permissions to compare with target. But it is also OK to just issue EMODPR with the target permissions.
Similarly in enclave, to avoid extra overhead of EMODPR, ocall, it'd help to keep track of EPCM permissions of each page. 
I updated with pseudo code to show possible implementation in enclave without tracking permissions.

Regardless of implementation, the mprotect interface should have worked.